build_resolvers/lib/src/build_asset_uri_resolver.dart

```diff
@@ -11,70 +12,83 @@ import 'package:analyzer/file_system/memory_file_system.dart';
 import 'package:analyzer/src/dart/analysis/driver.dart' show AnalysisDriver;
 import 'package:analyzer/src/generated/source.dart';
 import 'package:build/build.dart' show AssetId, BuildStep;
+import 'package:crypto/crypto.dart';
+import 'package:graphs/graphs.dart';
 import 'package:path/path.dart' as p;
 
 const _ignoredSchemes = ['dart', 'dart-ext'];
 
 class BuildAssetUriResolver extends UriResolver {
+  /// A cache of the directives for each Dart library.
+  ///
+  /// This is stored across builds and is only invalidated if we read a file and
+  /// see that it's content is different from what it was last time it was read.
   final _cachedAssetDependencies = <AssetId, Set<AssetId>>{};
-  final _cachedAssetContents = <AssetId, String>{};
+
+  /// A cache of the digest for each Dart asset.
+  ///
+  /// This is stored across builds and used to invalidate the values in
+  /// [_cachedAssetDependencies] only when the actual content of the library
+  /// changed.
+  final _cachedAssetDigests = <AssetId, Digest>{};
```

Comment:
Comparing digests might actually be slower than strings? Digests that are equal aren't identical, but Strings I think are canonicalized?

In practice I believe this actually won't matter because we keep the old digest instances around instead of bashing over with the new one if it doesn't change.

Reply:
> Strings I think are canonicalized?

Not always. If you read a file into a string twice then `first == second` will be true but `identical(first, second)` will not be true. I think the `identical` case would happen for either `Digest` or `String` for us anyway, since we cache things and should always give back the same instance, but I don't know if the `operator ==` for `Digest` takes advantage of that.


Either way, I don't think I can find any performance difference but I didn't add a stopwatch... I wouldn't be surprised to find that when files are changed the performance improvement we get from comparing digests more the offsets the potential optimizations in the equals case.

Reply:
The `==` operator for Digest uses `ListEquality` from package:collection which does do an instance check first.

I think this should be fine since we cache the instances.